homeassistant/components/control4/__init__.py

```diff
@@ -0,0 +1,229 @@
+"""The Control4 integration."""
+import asyncio
+import json
+import logging
+
+from aiohttp import client_exceptions
+from pyControl4.account import C4Account
+from pyControl4.director import C4Director
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    CONF_HOST,
+    CONF_PASSWORD,
+    CONF_SCAN_INTERVAL,
+    CONF_TOKEN,
+    CONF_USERNAME,
+)
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import device_registry as dr, entity
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import (
+    CONF_ACCOUNT,
+    CONF_CONFIG_LISTENER,
+    CONF_CONTROLLER_UNIQUE_ID,
+    CONF_DIRECTOR,
+    CONF_DIRECTOR_ALL_ITEMS,
+    CONF_DIRECTOR_MODEL,
+    CONF_DIRECTOR_SW_VERSION,
+    CONF_DIRECTOR_TOKEN_EXPIRATION,
+    CONF_LIGHT_COLD_START_TRANSITION_TIME,
+    CONF_LIGHT_TRANSITION_TIME,
+    DEFAULT_LIGHT_COLD_START_TRANSITION_TIME,
+    DEFAULT_LIGHT_TRANSITION_TIME,
+    DEFAULT_SCAN_INTERVAL,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["light"]
+
+
+async def async_setup(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Stub to allow setting up this component.
+
+    Configuration through YAML is not supported at this time.
+    """
+    hass.data.setdefault(DOMAIN, {})
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up Control4 from a config entry."""
+    entry_data = hass.data[DOMAIN].setdefault(entry.entry_id, {})
+
+    config = entry.data
+    account = C4Account(config[CONF_USERNAME], config[CONF_PASSWORD])
+    try:
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
```

Comment:
Would it be possible to distinguish about expired authentication and unable to reach the server?

Reply:
This step is retrieving an authentication token using the stored username and password, so I'll try to implement a check for invalid username/password.

Reply:
Would it be possible to not have to store user/pass and just keep tokens around ?

Reply:
I don't think so - even though the API says that account tokens never expire, I've found that sometimes they stop working, requiring re-authentication. 

The Control4 API is not publicly documented, so its safer to allow the integration to store the username/password and re-auth every time.